resources/type-system/inference.md

```diff
@@ -969,6 +969,23 @@ with respect to `L` under constraints `C0`
   - If for `i` in `0...m`, `Mi` is a subtype match for `Ni` with respect to `L`
   under constraints `Ci`.
 
+## Bound resolution
```

Comment:
This is already defined on page 144 of the latest spec at spec.dart.dev, consider just referencing that rather than duplicating?

Reply:
As I see it, there’s an important difference. The spec notion of “*T* is *T<sub>0</sub>* bounded” is a relation between types but not a function. For example, if `A` and `B` are type variables, the bound of `A` is `B`, and the bound of `B` is `C`, then `A` is considered to simultaneously be `A` bounded, `B` bounded, and `C` bounded.

The notion I’m defining here is a total function: considering the same example, the “bound resolution” of `A` is uniquely `C`. The way I use this notion later in the PR relies on it being a total function (a step in the “Method invocation inference” procedure is "Let `U_0` be the bound resolution of `T_0`").

The two notions are definitely related, and I guess we could define bound resolution in terms of "*T<sub>0</sub>* bounded" as follows:

> The bound resolution of *T* is the unique type *T<sub>0</sub>* such that *T<sub>0</sub>* is neither a type variable nor a union of the form *X&B*.

But then we would need to justify why such a type is known to exist and be unique for all *T*, and at that point it seems to me like defining it procedurally is both clearer and easier to reconcile with what the implementations do.

Reply:
I really don't like having this duplicated here, and I still think it is duplicated for all practical purposes.  I think what you are getting at is that the spec chooses to say that `T0` is `T0` bounded without the qualification that `T0` is neither a type variable nor an intersection.  I think all we need to do to make these definitions coincide is add that qualification - I see no reason that we want to say that a type variable `X` is `X bounded`.  I'd suggest fixing the spec on this rather than duplicating this here.  cc @eernstg what do you think?

Reply:
I've created a sketch of what this might look like here: https://github.com/dart-lang/language/pull/4013